pkg/collector/upgrade/versions.go

```diff
@@ -1,35 +1,31 @@
 package upgrade
 
 import (
+	"github.com/Masterminds/semver/v3"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/open-telemetry/opentelemetry-operator/api/v1alpha1"
 )
 
-// Version represents a known OpenTelemetry Collector version
-type Version struct {
-	// Tag represents the tag for this version, like: 0.0.1
-	Tag string
+type upgradeFunc func(cl client.Client, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error)
 
-	upgrade func(cl client.Client, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error)
-	next    *Version
+type otelcolVersion struct {
+	semver.Version
+	upgrade upgradeFunc
 }
 
 var (
-	v0_0_1 = Version{Tag: "0.0.1", upgrade: noop, next: &v0_0_2}
-	v0_0_2 = Version{Tag: "0.0.2", upgrade: noop, next: &v0_2_0}
-	v0_2_0 = Version{Tag: "0.2.0", upgrade: upgrade0_2_0}
-
-	// Latest represents the latest known version for the OpenTelemetry Collector
-	Latest = &v0_2_0
-
-	versions = map[string]Version{
-		v0_0_1.Tag: v0_0_1,
-		v0_0_2.Tag: v0_0_2,
-		v0_2_0.Tag: v0_2_0,
+	versions = []otelcolVersion{
+		{
+			Version: *semver.MustParse("0.2.10"),
+			upgrade: upgrade0_2_10,
+		},
+		{
+			Version: *semver.MustParse("0.9.0"),
+			upgrade: upgrade0_9_0,
+		},
 	}
-)
 
-func noop(cl client.Client, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error) {
-	return otelcol, nil
-}
+	// Latest represents the latest version that we need to upgrade. This is not necessarily the latest known version.
+	Latest = versions[len(versions)-1]
```

Comment:
Just wondering if this is necessary? i.e. what purpose does it serve. The important versions are the instance's current version, and the max version supported by the operator?

Reply:
We have a special condition in the logic where we would skip the upgrade if the version set in the CR is higher than the one we known, like in the case for downgrades. If I expose the whole map of versions, I can remove this latest, and calculate the latest on those two places.